ForceMergeActionTests.testToSteps failure

https://gradle-enterprise.elastic.co/s/uspzcasiyxcli

```
java.lang.AssertionError: 	
    Expected: <{"phase":"oLERN","action":"forcemerge","name":"forcemerge"} => {"phase":"oLERN","action":"forcemerge","name":"close-index"}>	
         but: was <{"phase":"oLERN","action":"forcemerge","name":"forcemerge"}>	
        at __randomizedtesting.SeedInfo.seed([6BBB6E98BBA45352:9A4D547F55401BF]:0)	
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)	
        at org.junit.Assert.assertThat(Assert.java:956)	
        at org.junit.Assert.assertThat(Assert.java:923)	
        at org.elasticsearch.xpack.core.ilm.ForceMergeActionTests.assertBestCompression(ForceMergeActionTests.java:93)	
        at org.elasticsearch.xpack.core.ilm.ForceMergeActionTests.testToSteps(ForceMergeActionTests.java:135)
```

Introduced by #49974

--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeAction.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeAction.java
@@ -118,31 +118,35 @@ public class ForceMergeAction implements LifecycleAction {
         Settings bestCompressionSettings = Settings.builder()
             .put(EngineConfig.INDEX_CODEC_SETTING.getKey(), CodecService.BEST_COMPRESSION_CODEC).build();
 
+        final boolean codecChange = codec != null && codec.equals(CodecService.BEST_COMPRESSION_CODEC);
+
         StepKey readOnlyKey = new StepKey(phase, NAME, ReadOnlyAction.NAME);
+
+        StepKey closeKey = new StepKey(phase, NAME, CloseIndexStep.NAME);
+        StepKey updateCompressionKey = new StepKey(phase, NAME, UpdateSettingsStep.NAME);
+        StepKey openKey = new StepKey(phase, NAME, OpenIndexStep.NAME);
+        StepKey waitForGreenIndexKey = new StepKey(phase, NAME, WaitForIndexColorStep.NAME);
+
         StepKey forceMergeKey = new StepKey(phase, NAME, ForceMergeStep.NAME);
         StepKey countKey = new StepKey(phase, NAME, SegmentCountStep.NAME);
 
-        StepKey closeKey = new StepKey(phase, NAME, CloseIndexStep.NAME);
-        StepKey openKey = new StepKey(phase, NAME, OpenIndexStep.NAME);
-        StepKey waitForGreenIndexKey = new StepKey(phase, NAME, WaitForIndexColorStep.NAME);
-        StepKey updateCompressionKey = new StepKey(phase, NAME, UpdateSettingsStep.NAME);
+        UpdateSettingsStep readOnlyStep =
+            new UpdateSettingsStep(readOnlyKey, codecChange ? closeKey : forceMergeKey, client, readOnlySettings);
 
-        UpdateSettingsStep readOnlyStep = new UpdateSettingsStep(readOnlyKey, forceMergeKey, client, readOnlySettings);
-        ForceMergeStep forceMergeStep = new ForceMergeStep(forceMergeKey, countKey, client, maxNumSegments);
         CloseIndexStep closeIndexStep = new CloseIndexStep(closeKey, updateCompressionKey, client);
-        ForceMergeStep forceMergeStepForBestCompression = new ForceMergeStep(forceMergeKey, closeKey, client, maxNumSegments);
         UpdateSettingsStep updateBestCompressionSettings = new UpdateSettingsStep(updateCompressionKey,
             openKey, client, bestCompressionSettings);
         OpenIndexStep openIndexStep = new OpenIndexStep(openKey, waitForGreenIndexKey, client);
         WaitForIndexColorStep waitForIndexGreenStep = new WaitForIndexColorStep(waitForGreenIndexKey,
             forceMergeKey, ClusterHealthStatus.GREEN);
+
+        ForceMergeStep forceMergeStep = new ForceMergeStep(forceMergeKey, countKey, client, maxNumSegments);
         SegmentCountStep segmentCountStep = new SegmentCountStep(countKey, nextStepKey, client, maxNumSegments);
 
         List<Step> mergeSteps = new ArrayList<>();
         mergeSteps.add(readOnlyStep);
 
-        if (codec != null && codec.equals(CodecService.BEST_COMPRESSION_CODEC)) {
-            mergeSteps.add(forceMergeStepForBestCompression);
+        if (codecChange) {
             mergeSteps.add(closeIndexStep);
             mergeSteps.add(updateBestCompressionSettings);
             mergeSteps.add(openIndexStep);

--- a/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/ForceMergeActionTests.java
+++ b/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/ForceMergeActionTests.java
@@ -7,6 +7,7 @@ package org.elasticsearch.xpack.core.ilm;
 
 import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.collect.Tuple;
 import org.elasticsearch.common.io.stream.Writeable.Reader;
 import org.elasticsearch.common.xcontent.DeprecationHandler;
 import org.elasticsearch.common.xcontent.XContentHelper;
@@ -19,7 +20,9 @@ import org.elasticsearch.xpack.core.ilm.Step.StepKey;
 
 import java.io.IOException;
 import java.util.List;
+import java.util.stream.Collectors;
 
+import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.equalTo;
 
 public class ForceMergeActionTests extends AbstractActionTestCase<ForceMergeAction> {
@@ -80,33 +83,31 @@ public class ForceMergeActionTests extends AbstractActionTestCase<ForceMergeActi
         StepKey nextStepKey = new StepKey(randomAlphaOfLength(10), randomAlphaOfLength(10), randomAlphaOfLength(10));
         List<Step> steps = instance.toSteps(null, phase, nextStepKey);
         assertNotNull(steps);
-        assertEquals(8, steps.size());
+        assertEquals(7, steps.size());
+        List<Tuple<StepKey, StepKey>> stepKeys = steps.stream()
+            .map(s -> new Tuple<>(s.getKey(), s.getNextStepKey()))
+            .collect(Collectors.toList());
+        StepKey readOnly = new StepKey(phase, ForceMergeAction.NAME, ReadOnlyAction.NAME);
+        StepKey closeIndex = new StepKey(phase, ForceMergeAction.NAME, CloseIndexStep.NAME);
+        StepKey updateCodec = new StepKey(phase, ForceMergeAction.NAME, UpdateSettingsStep.NAME);
+        StepKey openIndex = new StepKey(phase, ForceMergeAction.NAME, OpenIndexStep.NAME);
+        StepKey waitForGreen = new StepKey(phase, ForceMergeAction.NAME, WaitForIndexColorStep.NAME);
+        StepKey forceMerge = new StepKey(phase, ForceMergeAction.NAME, ForceMergeStep.NAME);
+        StepKey segmentCount = new StepKey(phase, ForceMergeAction.NAME, SegmentCountStep.NAME);
+        assertThat(stepKeys, contains(
+            new Tuple<>(readOnly, closeIndex),
+            new Tuple<>(closeIndex, updateCodec),
+            new Tuple<>(updateCodec, openIndex),
+            new Tuple<>(openIndex, waitForGreen),
+            new Tuple<>(waitForGreen, forceMerge),
+            new Tuple<>(forceMerge, segmentCount),
+            new Tuple<>(segmentCount, nextStepKey)));
+
         UpdateSettingsStep firstStep = (UpdateSettingsStep) steps.get(0);
-        ForceMergeStep secondStep = (ForceMergeStep) steps.get(1);
-        CloseIndexStep thirdStep = (CloseIndexStep) steps.get(2);
-        UpdateSettingsStep fourthStep = (UpdateSettingsStep) steps.get(3);
-        OpenIndexStep fifthStep = (OpenIndexStep) steps.get(4);
-        WaitForIndexColorStep sixthStep = (WaitForIndexColorStep) steps.get(5);
-        ForceMergeStep seventhStep = (ForceMergeStep) steps.get(6);
-        SegmentCountStep eighthStep = (SegmentCountStep) steps.get(7);
-        assertThat(firstStep.getKey(), equalTo(new StepKey(phase, ForceMergeAction.NAME, ReadOnlyAction.NAME)));
-        assertThat(firstStep.getNextStepKey(), equalTo(secondStep));
+        UpdateSettingsStep thirdStep = (UpdateSettingsStep) steps.get(2);
+
         assertTrue(IndexMetaData.INDEX_BLOCKS_WRITE_SETTING.get(firstStep.getSettings()));
-        assertThat(secondStep.getKey(), equalTo(new StepKey(phase, ForceMergeAction.NAME, ForceMergeAction.NAME)));
-        assertThat(secondStep.getNextStepKey(), equalTo(thirdStep));
-        assertThat(thirdStep.getKey(), equalTo(new StepKey(phase, ForceMergeAction.NAME, CloseIndexStep.NAME)));
-        assertThat(thirdStep.getNextStepKey(), equalTo(fourthStep.getKey()));
-        assertThat(fourthStep.getKey(), equalTo(new StepKey(phase, ForceMergeAction.NAME, UpdateSettingsStep.NAME)));
-        assertThat(fourthStep.getSettings().get(EngineConfig.INDEX_CODEC_SETTING.getKey()), equalTo(CodecService.BEST_COMPRESSION_CODEC));
-        assertThat(fourthStep.getNextStepKey(), equalTo(fifthStep.getKey()));
-        assertThat(fifthStep.getKey(), equalTo(new StepKey(phase, ForceMergeAction.NAME, OpenIndexStep.NAME)));
-        assertThat(fifthStep.getNextStepKey(), equalTo(sixthStep));
-        assertThat(sixthStep.getKey(), equalTo(new StepKey(phase, ForceMergeAction.NAME, WaitForIndexColorStep.NAME)));
-        assertThat(sixthStep.getNextStepKey(), equalTo(seventhStep));
-        assertThat(seventhStep.getKey(), equalTo(new StepKey(phase, ForceMergeAction.NAME, ForceMergeStep.NAME)));
-        assertThat(seventhStep.getNextStepKey(), equalTo(nextStepKey));
-        assertThat(eighthStep.getKey(), equalTo(new StepKey(phase, ForceMergeAction.NAME, SegmentCountStep.NAME)));
-        assertThat(eighthStep.getNextStepKey(), equalTo(nextStepKey));
+        assertThat(thirdStep.getSettings().get(EngineConfig.INDEX_CODEC_SETTING.getKey()), equalTo(CodecService.BEST_COMPRESSION_CODEC));
     }
 
     public void testMissingMaxNumSegments() throws IOException {
@@ -129,13 +130,11 @@ public class ForceMergeActionTests extends AbstractActionTestCase<ForceMergeActi
         assertThat(r.getMessage(), equalTo("unknown index codec: [DummyCompressingStoredFields]"));
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/51822")
     public void testToSteps() {
         ForceMergeAction instance = createTestInstance();
         if (instance.getCodec() != null && CodecService.BEST_COMPRESSION_CODEC.equals(instance.getCodec())) {
             assertBestCompression(instance);
-        }
-        else {
+        } else {
             assertNonBestCompression(instance);
         }
     }

